test/metabase/lib/util_test.cljc

```diff
@@ -0,0 +1,123 @@
+(ns metabase.lib.util-test
+  (:require [metabase.lib.util :as lib.util]
+            #?(:clj [clojure.test :as t]
+               :cljs [cljs.test :as t :include-macros true])))
+
+(t/deftest ^:parallel roll-query-test
+  (t/is (= {:database 1
+            :type     :pipeline
+            :stages   [{:source-table 2}
+                       {}
+                       {:filter [:= [:field 3 nil] "wow"]}]}
+           (lib.util/roll-query {:database 1
+                                 :query    {:source-query {:source-query {:source-table 2}}
+                                            :filter       [:= [:field 3 nil] "wow"]}}))))
+
+(t/deftest ^:parallel unroll-query-test
+  (t/is (= {:database 1
+            :type     :query
+            :query    {:source-query {:source-query {:source-table 2}}
+                       :filter       [:= [:field 3 nil] "wow"]}}
+           (lib.util/unroll-query {:database 1
+                                   :type     :pipeline
+                                   :stages   [{:source-table 2}
+                                              {}
```

Comment:
Should `unroll-query` be dropping empty stages? It seems like we might create extra stages that end up having nothing in them. I suppose they're harmless to the query processor?

On the other hand we wrap a native query with an empty stage elsewhere, and if it stays empty it should probably be dropped? Or am I fretting about human optics over things the QP actually cares about.

Tangentially, it's unclear to me what constitutes a "native" query in this scheme, which the UI will care about. Anything with a native first stage, even if we're adding filters to it or something?